File: tests/python/test_parse_map_data.py
"""
Tests for Python code that parses, visualizes, and describes map drawing
data from Qualtrics survey exports.

Why these matter: Same rationale as the R tests -- the researcher's analysis
pipeline must be able to ingest WKT strings, merge them with survey response
data, and do spatial analysis. Some researchers will use Python
(geopandas/shapely) instead of R (sf). Both paths must work.

The format is WKT (Well-Known Text), so parsing is one line:
    shapely.wkt.loads("POLYGON((-88.24 40.12, ...))")

Run with: python -m pytest tests/python/test_parse_map_data.py -v
Requires: pip install geopandas shapely matplotlib pytest
"""

import json
import pytest

import sys
import os

sys.path.insert(
    0, os.path.join(os.path.dirname(os.path.dirname(os.path.dirname(__file__))), "python")
)
from parse_map_data import (
    parse_coordinate_string,
    parse_qualtrics_map_data,
    compute_spatial_descriptives,
    plot_respondent_map,
    to_geodataframe,
)


# --- Fixture data (WKT format) ---

SAMPLE_CSV_ROWS = [
    {
        "ResponseID": "R_abc123",
        "MapDrawing": "POLYGON((-88.2434 40.1164, -88.2334 40.1064, -88.2234 40.1164, -88.2434 40.1164))",
        "zoom": 14,
        "Q1_community_attachment": 4,
        "Q2_years_in_neighborhood": 10,
        "MapAssignments": '{"overlayCondition":"ward_boundary","showTraffic":false}',
    },
    {
        "ResponseID": "R_def456",
        "MapDrawing": "MULTIPOLYGON(((-88.25 40.12, -88.23 40.10, -88.21 40.12, -88.25 40.12)), ((-88.30 40.15, -88.28 40.13, -88.26 40.15, -88.30 40.15)))",
        "zoom": 12,
        "Q1_community_attachment": 5,
        "Q2_years_in_neighborhood": 2,
        "MapAssignments": '{"overlayCondition":"census_tract","showTraffic":true}',
    },
    {
        "ResponseID": "R_ghi789",
        "MapDrawing": "",
        "zoom": 14,
        "Q1_community_attachment": 3,
        "Q2_years_in_neighborhood": 7,
        "MapAssignments": '{"overlayCondition":"control","showTraffic":false}',
    },
]


# --- Parsing tests ---


class TestParseCoordinateString:
    def test_one_polygon(self):
        wkt = "POLYGON((-88.2434 40.1164, -88.2334 40.1064, -88.2234 40.1164, -88.2434 40.1164))"
        result = parse_coordinate_string(wkt)
        assert len(result) == 1
        assert result[0].geom_type == "Polygon"

    def test_multiple_polygons(self):
        wkt = "MULTIPOLYGON(((-88.25 40.12, -88.23 40.10, -88.21 40.12, -88.25 40.12)), ((-88.30 40.15, -88.28 40.13, -88.26 40.15, -88.30 40.15)))"
        result = parse_coordinate_string(wkt)
        assert len(result) == 2

    def test_empty_string(self):
        result = parse_coordinate_string("")
        assert len(result) == 0

    def test_longitude_first(self):
        """WKT uses x y = longitude latitude order."""
        wkt = "POLYGON((-88.2434 40.1164, -88.2334 40.1064, -88.2234 40.1164, -88.2434 40.1164))"
        result = parse_coordinate_string(wkt)
        centroid = result[0].centroid
        assert centroid.x < -80  # longitude
        assert centroid.y > 30  # latitude

    def test_standard_wkt_no_custom_parser_needed(self):
        """shapely.wkt.loads handles it directly -- no custom code."""
        from shapely import wkt as shapely_wkt
        wkt = "POLYGON((-88.2434 40.1164, -88.2334 40.1064, -88.2234 40.1164, -88.2434 40.1164))"
        geom = shapely_wkt.loads(wkt)
        assert geom.geom_type == "Polygon"
        assert not geom.is_empty


# --- DataFrame conversion ---


class TestParseQualtricsCsv:
    def test_converts_to_geodataframe(self):
        import pandas as pd

        df = pd.DataFrame(SAMPLE_CSV_ROWS)
        gdf = to_geodataframe(df, coord_col="MapDrawing")
        assert hasattr(gdf, "geometry")
        assert len(gdf) == 3

    def test_preserves_survey_columns(self):
        import pandas as pd

        df = pd.DataFrame(SAMPLE_CSV_ROWS)
        gdf = to_geodataframe(df, coord_col="MapDrawing")
        assert "Q1_community_attachment" in gdf.columns
        assert "Q2_years_in_neighborhood" in gdf.columns
        assert "ResponseID" in gdf.columns

    def test_empty_drawing_produces_empty_geometry(self):
        """Respondents who did not draw should remain in the data."""
        import pandas as pd

        df = pd.DataFrame(SAMPLE_CSV_ROWS)
        gdf = to_geodataframe(df, coord_col="MapDrawing")
        assert gdf.geometry.iloc[2].is_empty
        assert gdf.iloc[2]["Q1_community_attachment"] == 3

    def test_crs_is_wgs84(self):
        import pandas as pd

        df = pd.DataFrame(SAMPLE_CSV_ROWS)
        gdf = to_geodataframe(df, coord_col="MapDrawing")
        assert gdf.crs.to_epsg() == 4326


# --- Assignment parsing ---


class TestAssignmentParsing:
    def test_unpacks_json_assignments(self):
        import pandas as pd

        df = pd.DataFrame(SAMPLE_CSV_ROWS)
        gdf = parse_qualtrics_map_data(
            df, coord_col="MapDrawing", assignments_col="MapAssignments"
        )
        assert "overlayCondition" in gdf.columns
        assert gdf.iloc[0]["overlayCondition"] == "ward_boundary"
        assert gdf.iloc[2]["overlayCondition"] == "control"

    def test_assignments_usable_as_categorical_variable(self):
        """Researcher should be able to group by overlay condition."""
        import pandas as pd

        df = pd.DataFrame(SAMPLE_CSV_ROWS)
        gdf = parse_qualtrics_map_data(
            df, coord_col="MapDrawing", assignments_col="MapAssignments"
        )
        grouped = gdf.groupby("overlayCondition")["Q1_community_attachment"].mean()
        assert len(grouped) == 3


# --- Spatial descriptives ---


class TestSpatialDescriptives:
    def test_computes_area(self):
        import pandas as pd

        df = pd.DataFrame(SAMPLE_CSV_ROWS)
        gdf = to_geodataframe(df, coord_col="MapDrawing")
        desc = compute_spatial_descriptives(gdf)
        assert desc["area_m2"].iloc[0] > 0

    def test_computes_centroid(self):
        import pandas as pd

        df = pd.DataFrame(SAMPLE_CSV_ROWS)
        gdf = to_geodataframe(df, coord_col="MapDrawing")
        desc = compute_spatial_descriptives(gdf)
        assert "centroid_lat" in desc.columns
        assert "centroid_lng" in desc.columns
        assert desc["centroid_lat"].iloc[0] > 30

    def test_empty_geometry_gets_nan_descriptives(self):
        import pandas as pd

        df = pd.DataFrame(SAMPLE_CSV_ROWS)
        gdf = to_geodataframe(df, coord_col="MapDrawing")
        desc = compute_spatial_descriptives(gdf)
        import math

        assert math.isnan(desc["area_m2"].iloc[2]) or desc["area_m2"].iloc[2] == 0


# --- Visualization ---


class TestVisualization:
    def test_plot_respondent_returns_figure(self):
        import pandas as pd

        df = pd.DataFrame(SAMPLE_CSV_ROWS)
        gdf = to_geodataframe(df, coord_col="MapDrawing")
        fig = plot_respondent_map(gdf, respondent_id="R_abc123")
        assert fig is not None

File: tests/R/test-parse-map-data.R
# Tests for R code that parses, visualizes, and describes map drawing data
# from Qualtrics survey exports.
#
# Why these matter: The researcher's analysis pipeline starts here. They
# download a CSV from Qualtrics with one row per respondent. One column
# contains a WKT geometry string from the map drawing. Other columns contain
# survey responses (possibly from randomized questions, conjoint experiments,
# etc.). The researcher needs to:
#   1. Parse the WKT string into spatial objects (sf)
#   2. Merge spatial data with survey response data for joint analysis
#   3. Visualize what respondents drew
#   4. Compute spatial descriptives (area, centroid, overlap)
#   5. Use these as outcomes or treatments in statistical models
#
# WKT is the standard geometry format for tabular data. sf::st_as_sfc()
# parses it natively -- no custom parser needed.
#
# Run with: Rscript -e 'testthat::test_file("tests/R/test-parse-map-data.R")'

library(testthat)
library(sf)

.project_root <- if (file.exists("R/parse_map_data.R")) {
  "."
} else if (file.exists("../../R/parse_map_data.R")) {
  "../.."
} else {
  stop("Cannot find R/parse_map_data.R -- run from project root or tests/R/")
}
source(file.path(.project_root, "R", "parse_map_data.R"))

# --- Fixture data ---
# Simulates what a researcher would see in their Qualtrics CSV export.
# MapDrawing column now contains WKT strings.
sample_qualtrics_export <- data.frame(
  ResponseID = c("R_abc123", "R_def456", "R_ghi789"),
  MapDrawing = c(
    # Respondent 1: one triangle near Champaign, IL
    "POLYGON((-88.2434 40.1164, -88.2334 40.1064, -88.2234 40.1164, -88.2434 40.1164))",
    # Respondent 2: two polygons
    "MULTIPOLYGON(((-88.25 40.12, -88.23 40.10, -88.21 40.12, -88.25 40.12)), ((-88.30 40.15, -88.28 40.13, -88.26 40.15, -88.30 40.15)))",
    # Respondent 3: drew nothing (clicked Done immediately)
    ""
  ),
  zoom = c(14, 12, 14),
  Q1_community_attachment = c(4, 5, 3),
  Q2_years_in_neighborhood = c(10, 2, 7),
  MapAssignments = c(
    '{"overlayCondition":"ward_boundary","showTraffic":false}',
    '{"overlayCondition":"census_tract","showTraffic":true}',
    '{"overlayCondition":"control","showTraffic":false}'
  ),
  stringsAsFactors = FALSE
)


# --- Parsing tests ---

test_that("parse_coordinate_string converts POLYGON WKT to an sf polygon", {
  wkt <- "POLYGON((-88.2434 40.1164, -88.2334 40.1064, -88.2234 40.1164, -88.2434 40.1164))"
  result <- parse_coordinate_string(wkt)

  expect_s3_class(result, "sfc")
  expect_equal(length(result), 1)
  expect_true(sf::st_is(result[[1]], "POLYGON"))
})

test_that("parse_coordinate_string handles MULTIPOLYGON WKT", {
  wkt <- "MULTIPOLYGON(((-88.25 40.12, -88.23 40.10, -88.21 40.12, -88.25 40.12)), ((-88.30 40.15, -88.28 40.13, -88.26 40.15, -88.30 40.15)))"
  result <- parse_coordinate_string(wkt)
  expect_equal(length(result), 2)
})

test_that("parse_coordinate_string returns empty geometry for empty string", {
  result <- parse_coordinate_string("")
  expect_equal(length(result), 0)
})

test_that("parsed coordinates are in WGS84 (EPSG:4326)", {
  wkt <- "POLYGON((-88.2434 40.1164, -88.2334 40.1064, -88.2234 40.1164, -88.2434 40.1164))"
  result <- parse_coordinate_string(wkt)
  crs <- sf::st_crs(result)
  expect_equal(crs$epsg, 4326L)
})

test_that("WKT uses longitude-first order, parsed correctly by sf", {
  # WKT coordinates are x y = longitude latitude.
  # sf reads them correctly: x -> easting (lon), y -> northing (lat).
  wkt <- "POLYGON((-88.2434 40.1164, -88.2334 40.1064, -88.2234 40.1164, -88.2434 40.1164))"
  result <- parse_coordinate_string(wkt)
  bbox <- sf::st_bbox(result)
  # xmin/xmax should be longitude (negative, around -88)
  expect_true(bbox["xmin"] < -80)
  # ymin/ymax should be latitude (positive, around 40)
  expect_true(bbox["ymin"] > 30)
})

test_that("legacy format (pre-WKT) is still parseable for old data", {
  # Data collected before April 2026 used "lon lat,lon lat;lon lat" format.
  # The parser auto-detects and handles both.
  legacy <- "-88.2434 40.1164,-88.2334 40.1064,-88.2234 40.1164"
  result <- parse_coordinate_string(legacy)
  expect_equal(length(result), 1)
  expect_true(sf::st_is(result[[1]], "POLYGON"))
  bbox <- sf::st_bbox(result)
  expect_true(bbox["xmin"] < -80)
})

test_that("legacy format with multiple polygons is parseable", {
  legacy <- "-88.25 40.12,-88.23 40.10,-88.21 40.12;-88.30 40.15,-88.28 40.13,-88.26 40.15"
  result <- parse_coordinate_string(legacy)
  expect_equal(length(result), 2)
})

test_that("standard WKT is readable without custom parsing code", {
  # The whole point of WKT: sf::st_as_sfc handles it natively.
  # No need to source any custom R functions.
  wkt <- "POLYGON((-88.2434 40.1164, -88.2334 40.1064, -88.2234 40.1164, -88.2434 40.1164))"
  result <- sf::st_as_sfc(wkt, crs = 4326)
  expect_s3_class(result, "sfc")
  expect_equal(length(result), 1)
  expect_true(sf::st_is(result[[1]], "POLYGON"))
})


# --- Data frame conversion tests ---

test_that("parse_qualtrics_map_data converts a data frame to sf", {
  result <- parse_qualtrics_map_data(sample_qualtrics_export,
    coord_col = "MapDrawing"
  )

  expect_s3_class(result, "sf")
  expect_equal(nrow(result), 3)
  expect_true("Q1_community_attachment" %in% names(result))
  expect_true("Q2_years_in_neighborhood" %in% names(result))
  expect_true("ResponseID" %in% names(result))
})

test_that("respondents with no drawing get empty geometry, not NA or error", {
  # Dropping respondents who did not draw would bias the sample.
  result <- parse_qualtrics_map_data(sample_qualtrics_export,
    coord_col = "MapDrawing"
  )
  expect_true(sf::st_is_empty(result$geometry[3]))
  expect_equal(result$Q1_community_attachment[3], 3)
})


# --- Experimental assignment parsing ---

test_that("parse_assignments extracts JSON assignment columns", {
  result <- parse_qualtrics_map_data(sample_qualtrics_export,
    coord_col = "MapDrawing",
    assignments_col = "MapAssignments"
  )
  expect_true("overlayCondition" %in% names(result))
  expect_true("showTraffic" %in% names(result))
  expect_equal(result$overlayCondition[1], "ward_boundary")
  expect_equal(result$overlayCondition[2], "census_tract")
  expect_equal(result$overlayCondition[3], "control")
})

test_that("treatment assignments are available for use as factors in models", {
  result <- parse_qualtrics_map_data(sample_qualtrics_export,
    coord_col = "MapDrawing",
    assignments_col = "MapAssignments"
  )
  expect_no_error({
    lm(Q1_community_attachment ~ overlayCondition, data = result)
  })
})


# --- Spatial descriptives ---

test_that("compute_spatial_descriptives returns area and centroid", {
  result <- parse_qualtrics_map_data(sample_qualtrics_export,
    coord_col = "MapDrawing"
  )
  descriptives <- compute_spatial_descriptives(result)

  expect_true("area_m2" %in% names(descriptives))
  expect_true("centroid_lat" %in% names(descriptives))
  expect_true("centroid_lng" %in% names(descriptives))
  expect_true(descriptives$area_m2[1] > 0)
  expect_true(is.na(descriptives$area_m2[3]) || descriptives$area_m2[3] == 0)
})

test_that("spatial descriptives can be merged with survey data for analysis", {
  result <- parse_qualtrics_map_data(sample_qualtrics_export,
    coord_col = "MapDrawing"
  )
  descriptives <- compute_spatial_descriptives(result)

  merged <- cbind(sf::st_drop_geometry(result), descriptives)
  expect_true("area_m2" %in% names(merged))
  expect_true("Q1_community_attachment" %in% names(merged))
  expect_equal(nrow(merged), 3)
})


# --- Visualization ---

test_that("plot_respondent_map returns a ggplot object", {
  result <- parse_qualtrics_map_data(sample_qualtrics_export,
    coord_col = "MapDrawing"
  )
  p <- plot_respondent_map(result, respondent_id = "R_abc123")
  expect_s3_class(p, "ggplot")
})

test_that("plot_all_maps returns a ggplot showing all respondents", {
  result <- parse_qualtrics_map_data(sample_qualtrics_export,
    coord_col = "MapDrawing"
  )
  p <- plot_all_maps(result)
  expect_s3_class(p, "ggplot")
})
